coral-hive/src/main/java/com/linkedin/coral/hive/hive2rel/functions/StaticHiveFunctionRegistry.java

```diff
@@ -501,6 +518,31 @@ public static void createAddUserDefinedFunction(String functionName, SqlReturnTy
     addFunctionEntry(functionName, createCalciteUDF(functionName, returnTypeInference, operandTypeChecker));
   }
 
+  /**
+   * Adds the generic UDTF, which is almost same as how we register for LinkedIn UDFs except that we need to register
+   * the return field names in `UDTF_RETURN_FIELD_NAME_MAP`
+   */
+  public static void createAddUserDefinedTableFunction(String functionName, ImmutableList<String> returnFieldNames,
+      ImmutableList<?> returnFieldTypes, SqlOperandTypeChecker operandTypeChecker) {
+    Preconditions.checkArgument(!returnFieldTypes.isEmpty() && returnFieldTypes.size() == returnFieldNames.size()
+        && (returnFieldTypes.stream().allMatch(type -> type instanceof SqlTypeName)
+            || returnFieldTypes.stream().allMatch(type -> type instanceof SqlReturnTypeInference)));
```

Comment:
just curious, why restricting return types to be either all `SqlTypeName` or `SqlReturnTypeInference` as opposed to have a mix of both?

Reply:
Because we need to call either `HiveReturnTypes.rowOf(ImmutableList<String> fieldNames, ImmutableList<SqlTypeName> types)` or `HiveReturnTypes.rowOfInference(ImmutableList<String> fieldNames, ImmutableList<SqlReturnTypeInference> types)` to get the return type.

We need 2 methods because `HiveReturnTypes.rowOfInference` is useful to create row types whose fields are complex types and hence cannot be represented through `SqlTypeName` to be used in `HiveReturnTypes.rowOf`